src/main/scala/zio/prelude/NonEmptySet.scala

```diff
@@ -0,0 +1,129 @@
+package zio.prelude
+
+import scala.language.implicitConversions
+
+sealed abstract case class NonEmptySet[A] private (elem: A, others: Set[A]) extends (A => Boolean) {
+  def toSet: Set[A]                   = others + elem
+  def toNonEmptyList: NonEmptyList[A] = NonEmptyList.fromIterable(elem, others)
+  def contains(elem0: A): Boolean     = toSet.contains(elem0)
+  def add(elem0: A): NonEmptySet[A]   = NonEmptySet.fromSet(elem0, toSet)
+  def remove(elem0: A): Set[A]        = toSet - elem0
+  def size: Int                       = others.size + 1
+
+  /**
+   * Flattens a `NonEmptySet` of `NonEmptySet` values into a single
+   * `NonEmptySet`.
+   */
+  final def flatten[B](implicit ev: A <:< NonEmptySet[B]): NonEmptySet[B] =
+    others.foldLeft(ev(elem)) { (b, a) =>
+      NonEmptySet.union(b, ev(a))
+    }
+
+  override def apply(elem0: A): Boolean = contains(elem0)
+
+  override def hashCode(): Int = toSet.hashCode()
+
+  override def equals(that: Any): Boolean = that match {
+    case that: AnyRef if this.eq(that) => true
+    case that: NonEmptySet[A]          => this.toSet == that.toSet
+    case _                             => false
+  }
+
+  override def toString: String = s"NonEmpty$toSet"
+}
+
+object NonEmptySet {
+  private def apply[A](elem: A, others: Set[A]): NonEmptySet[A] = new NonEmptySet(elem, others - elem) {}
+  def apply[A](elem: A, others: A*): NonEmptySet[A]             = apply(elem, others.toSet)
+
+  def fromSet[A](elem: A, others: Set[A]): NonEmptySet[A]   = apply(elem, others - elem)
+  def fromSetOption[A](set: Set[A]): Option[NonEmptySet[A]] = set.headOption.map(fromSet(_, set.tail))
+
+  /**
+   * Constructs a `NonEmptySet` from an element and `Iterable`.
+   */
+  def fromIterable[A](head: A, tail: Iterable[A]): NonEmptySet[A] =
+    fromSet(head, tail.toSet)
+
+  /**
+   * Constructs a `NonEmptySet` from an `Iterable` or `None` otherwise.
+   */
+  def fromIterableOption[A](iterable: Iterable[A]): Option[NonEmptySet[A]] =
+    iterable.headOption.fold(None: Option[NonEmptySet[A]])(h => Some(fromIterable(h, iterable.tail)))
+
+  /**
+   * Constructs a `NonEmptySet` with the specified single value.
+   */
+  def single[A](head: A): NonEmptySet[A] =
+    NonEmptySet(head)
+
+  def union[A](l: NonEmptySet[A], r: Set[A]): NonEmptySet[A] =
```

Comment:
I think these methods should all work with `set union nonEmptySet` with the implicit conversion. I think maybe the only thing we need here is implicit syntax on `Set` for when the right hand side is a `NonEmptySet`, because then `union` is guaranteed to return a `NonEmptySet`.